module/zfs/zvol.c

```diff
@@ -1400,9 +1414,19 @@ zvol_open(struct block_device *bdev, fmode_t flag)
 			goto out_mutex;
 	}
 
-	if ((flag & FMODE_WRITE) && (zv->zv_flags & ZVOL_RDONLY)) {
-		error = -EROFS;
-		goto out_open_count;
+	if (flag & FMODE_WRITE) {
+		if (zv->zv_flags & ZVOL_RDONLY) {
+			error = -EROFS;
+			goto out_open_count;
+		}
+		if (!(zv->zv_flags & ZVOL_OPENED_WR)) {
+			/*
+			 * The first time the zvol is opened as writable we
+			 * need to open a ZIL.
+			 */
+			zv->zv_flags |= ZVOL_OPENED_WR;
+			zv->zv_zilog = zil_open(zv->zv_objset, zvol_get_data);
```

Comment:
I think it would be a good idea to add an assertion to `zvol_write()` and `zvol_discard()` that `ZVOL_OPENED_WR` is set (or that `zv->zv_zilog != NULL`) since they may call `zil_commit()`.